mlir-sycl/lib/Dialect/Analysis/AliasAnalysis.cpp

```diff
@@ -0,0 +1,125 @@
+//===- AliasAnalysis.cpp - Alias Analysis for the SYCL MLIR dialect -------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SYCL/Analysis/AliasAnalysis.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOpsDialect.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "alias-analysis"
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+// Return true if the value \p val is a function argument, and false otherwise.
+static bool isFuncArg(Value val) {
+  auto blockArg = val.dyn_cast<BlockArgument>();
+  if (!blockArg)
+    return false;
+
+  return isa_and_nonnull<FunctionOpInterface>(
+      blockArg.getOwner()->getParentOp());
+}
+
+// Return true if the value \p val is a function argument that has the
+// 'local_alias_analysis.restrict' attribute, and false otherwise.
+static bool isRestrict(Value val) {
+  if (!isFuncArg(val))
+    return false;
+
+  auto blockArg = val.cast<BlockArgument>();
+  auto func = cast<FunctionOpInterface>(blockArg.getOwner()->getParentOp());
+  return !!func.getArgAttr(blockArg.getArgNumber(),
+                           "local_alias_analysis.restrict");
+}
+
+// Return true if the operation \p op belongs to the SYCL MLIR dialect.
+static bool isSYCLOperation(Operation *op) {
+  if (!op || !op->getDialect())
+    return false;
+  return op->getDialect()->getNamespace() ==
+         sycl::SYCLDialect::getDialectNamespace();
+}
```

Comment:
```suggestion
// Return true if the operation \p op belongs to the SYCL MLIR dialect.
static bool isSYCLOperation(Operation *op) {
  if (!op || !op->getDialect())
    return false;
  return isa<SYCLDialect>(op->getDialect()->getNamespace());
}
```
Can we also move this to `mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.h` and make it `inline` there?